### PR TITLE
STM32G4x/STM32L4Rx/STM32L4Sx: Fix sys clock frequency and boost voltage for highest clock frequencies

### LINKED ISSUE #22245: STM32G4xx: Wrong SystemCoreClock variable
**Description**
Because of a bug in STM32Cube LL, the `SystemCoreClock` variable is set wrong if the actual system frequency is >80 MHz.

More details are described in a related (but already closed) PR:
https://github.com/zephyrproject-rtos/hal_stm32/pull/34#issuecomment-578881473

I raised this issue to track the bug, as I can't raise an issue in the original HAL repository.

**Impact**
Boards with STM32G431 are hardly useable with high clock frequencies, as e.g. UART baud rates are get configured wrong.
Preliminary workaround could be to reduce the system frequency to <80 MHz.

### PR DESCRIPTION
**Pull request by FRASTM**
This patch enables the main regulator boost voltage for sys clock highest frequency.

For STM32G4xx series, boost mode when sys clock is higher than 150MHz
For STM32L4Rx or STM32L4Sx series, boost mode when sys clock is higher than 80MHz

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/22245

Signed-off-by: Francois Ramu <francois.ramu@st.com>

### CODE REVIEW

File: drivers/clock_control/clock_stm32_ll_common.c [hunk 1]
```diff
@@ -305,6 +305,13 @@ static int stm32_clock_control_init(struct device *dev)
 #ifdef CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL
 	LL_UTILS_PLLInitTypeDef s_PLLInitStruct;
 
+#ifdef CONFIG_SOC_SERIES_STM32G4X
+	/* set power boost mode */
```
Comment (erwango) [inline drivers/clock_control/clock_stm32_ll_common.c]: Let's move this at the very end of the function.

File: drivers/serial/uart_stm32.c [hunk 2]
```diff
@@ -53,6 +53,12 @@ static inline void uart_stm32_set_baudrate(struct device *dev, u32_t baud_rate)
 		return;
 	}
 
+#if defined(CONFIG_LPUART_1) \
+	&& defined(CONFIG_SOC_SERIES_STM32G4X) \
+	&& defined(CONFIG_UART_CONSOLE)
```
Comment (erwango) [inline drivers/serial/uart_stm32.c]: I wonder why the dependency on `CONFIG_UART_CONSOLE`.

File: drivers/serial/uart_stm32.c [hunk 3]
```diff
@@ -665,6 +671,13 @@ static int uart_stm32_init(struct device *dev)
 
 	LL_USART_Disable(UartInstance);
 
+#if defined(CONFIG_LPUART_1) \
+	&& defined(CONFIG_SOC_SERIES_STM32G4X) \
+	&& defined(CONFIG_UART_CONSOLE)
+	/* the clock source for LPUART console is set to HSI */
+	LL_APB1_GRP2_EnableClock(LL_APB1_GRP2_PERIPH_LPUART1);
```
Comment (erwango) [inline drivers/serial/uart_stm32.c]: Should be done by stm32_clock_control_on.

File: drivers/serial/uart_stm32.c [hunk 4]
```diff
@@ -665,6 +671,13 @@ static int uart_stm32_init(struct device *dev)
...
+	LL_RCC_SetLPUARTClockSource(LL_RCC_LPUART1_CLKSOURCE_HSI);
```
Comment (erwango) [inline drivers/serial/uart_stm32.c]: I'm a bit concerned: by doing it that way, we're introducing hard coded configuration which doesn't match default SoC configuration.
We should be able to use default source clck configuration, which doesn't require all this.
Can you elaborate why default PCLK source for LPUART doesn't work ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/clock_control/clock_stm32_ll_common.c [hunk 5]
```diff
@@ -305,6 +305,13 @@ static int stm32_clock_control_init(struct device *dev)
...
+	if (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC >= 170000000) {
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/clock_control/clock_stm32_ll_common.c]: ok

File: drivers/clock_control/clock_stm32_ll_common.c (hunk 1 above)
Comment (FRASTM) [reply drivers/clock_control/clock_stm32_ll_common.c]: This voltage regulator must be set to enable range1 Boost mode **before** enabling the PLL.
So in any case, before the `LL_PLL_ConfigSystemClock_HSI()` function in the stm32_clock_control_init

File: drivers/serial/uart_stm32.c (hunk 2 above)
Comment (FRASTM) [reply drivers/serial/uart_stm32.c]: this fix will output correctly on console on nucleo_g431rb (where console is assigned to LPUART1). Other use cases are not considered (tested).

File: drivers/serial/uart_stm32.c (hunk 3 above)
Comment (FRASTM) [reply drivers/serial/uart_stm32.c]: already done in stm32_clock_control_on, so removed there

File: soc/arm/st_stm32/stm32g4/soc.c [hunk 6]
```diff
@@ -41,6 +41,10 @@ static int stm32g4_init(struct device *arg)
 	/* At reset, system core clock is set to 16 MHz from HSI */
 	SystemCoreClock = 16000000;
 
+#ifdef CONFIG_DEBUG
+	LL_DBGMCU_EnableDBGSleepMode();
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply soc/arm/st_stm32/stm32g4/soc.c]: ok

File: drivers/serial/uart_stm32.c (hunk 4 above)
Comment (FRASTM) [reply drivers/serial/uart_stm32.c]: requires register access to g431 to investigate further

**Comment by FRASTM** [comment]
power boost does not impact clock control but only valid for stm32g4.

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by rosterloh** [comment]
Not sure if this is helpful to this issue but I needed to change stm32g4xx_ll_utils.c in the stm32 hal in order to get the correct baud rate. 
`
status = UTILS_EnablePLLAndSwitchSystem(pllfreq, UTILS_ClkInitStruct);
`
Needs to be added to line 335. After AHBCLK divider is set back to 1 the system clock variable is not updated so while it's running at the correct speed the functions the request the clock value are returning half the actual value.

File: drivers/clock_control/clock_stm32g4.c [hunk 7]
```diff
@@ -29,6 +29,11 @@ void config_pll_init(LL_UTILS_PLLInitTypeDef *pllinit)
 	pllinit->PLLM = pllm(CONFIG_CLOCK_STM32_PLL_M_DIVISOR);
 	pllinit->PLLN = CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER;
 	pllinit->PLLR = pllr(CONFIG_CLOCK_STM32_PLL_R_DIVISOR);
+
+	/* set power boost mode */
+	if (sys_clock_hw_cycles_per_sec() >= 170000000) {
```
Comment (rosterloh) [inline drivers/clock_control/clock_stm32g4.c]: Shouldn't this be 150000000?

**Comment by martinjaeger** [comment]
So this PR attempts to fix also #21681, which was not mentioned yet.

I agree with @rosterloh that there is a bug in the STM LL driver, see also [his PR](https://github.com/zephyrproject-rtos/hal_stm32/pull/36). I think that's the root cause of the wrong UART baud rate. As far as I could see, the MCU actually runs at the specified speed, but only SystemCoreClock variable is set wrong in LL drivers. Did quite a lot of debugging looking at the actual register contents.

File: drivers/clock_control/clock_stm32g4.c (hunk 7 above)
Comment (FRASTM) [reply drivers/clock_control/clock_stm32g4.c]: yes, 150MHz

**Comment by FRASTM** [comment]
stm32g4xx: set power boost mode for sys clock greater than 150MHz

**Review by erwango — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Comment by FRASTM** [comment]
updating west.yml after merge of the https://github.com/zephyrproject-rtos/hal_stm32/pull/37

**Review by erwango — APPROVED** [review]

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please update the PR summary to include which issues it fixes. i.e., "Fixes #NNNN"

File: drivers/clock_control/clock_stm32l4_wb.c [hunk 8]
```diff
@@ -30,6 +30,12 @@ void config_pll_init(LL_UTILS_PLLInitTypeDef *pllinit)
 	pllinit->PLLM = pllm(CONFIG_CLOCK_STM32_PLL_M_DIVISOR);
 	pllinit->PLLN = CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER;
 	pllinit->PLLR = pllr(CONFIG_CLOCK_STM32_PLL_R_DIVISOR);
+#ifdef PWR_CR5_R1MODE
+	/* set power boost mode for sys clock greater than 80MHz */
+	if (sys_clock_hw_cycles_per_sec() >= 80000000) {
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/clock_control/clock_stm32l4_wb.c]: changed to MHZ(80)

File: drivers/clock_control/clock_stm32g4.c [hunk 9]
```diff
@@ -29,6 +29,11 @@ void config_pll_init(LL_UTILS_PLLInitTypeDef *pllinit)
 	pllinit->PLLM = pllm(CONFIG_CLOCK_STM32_PLL_M_DIVISOR);
 	pllinit->PLLN = CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER;
 	pllinit->PLLR = pllr(CONFIG_CLOCK_STM32_PLL_R_DIVISOR);
+
+	/* set power boost mode for sys clock greater than 150MHz */
+	if (sys_clock_hw_cycles_per_sec() >= 150000000) {
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/clock_control/clock_stm32g4.c]: changed to MHZ(150)

**Comment by FRASTM** [comment]
change to MHZ(xxx)
and describe PR : Fixes https://github.com/zephyrproject-rtos/zephyr/issues/22245

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**